### PR TITLE
Update Architecture to unknown machine

### LINKED ISSUE #96978: [RISC-V] Optimizing assemblies for size failed
### Description

On 8.0 release runtime can be built with the following command without any issues:
```
cd runtime
./build.sh --ci -c Release --cross --arch riscv64
```
However there is an error on the upcoming 9.x:
```
error NETSDK1144: Optimizing assemblies for size failed. Optimization can be disabled by setting the PublishTrimmed property to false. [/__w/dotnet_riscv/dotnet_riscv/runtime/src/coreclr/tools/aot/crossgen2/crossgen2_publish.csproj]
```

@ashaurtaev, I'm not totally sure that this is related to https://github.com/dotnet/runtime/pull/95188 but I were able to pass building process by using Mono by default. Sorry, not an expect of Dotnet and I don't know the relationship between Mono and Crossgen2, but this might be a kind of regression.

### Reproduction Steps

```
./build.sh --ci -c Release --cross --arch riscv64
```

### Expected behavior

successful build process

### Actual behavior

error NETSDK1144

### Regression?

_No response_

### Known Workarounds

workaround by patching [eng/Subsets.props](https://github.com/dotnet/runtime/blob/main/eng/Subsets.props), but that doesn’t seem correct.
```patch
--- a/eng/Subsets.props
+++ b/eng/Subsets.props
@@ -29,6 +29,7 @@
     <DefaultPrimaryRuntimeFlavor>CoreCLR</DefaultPrimaryRuntimeFlavor>
     <DefaultPrimaryRuntimeFlavor Condition="'$(TargetArchitecture)' == 'armv6'">Mono</DefaultPrimaryRuntimeFlavor>
     <DefaultPrimaryRuntimeFlavor Condition="'$(TargetArchitecture)' == 'ppc64le'">Mono</DefaultPrimaryRuntimeFlavor>
+    <DefaultPrimaryRuntimeFlavor Condition="'$(TargetArchitecture)' == 'riscv64'">Mono</DefaultPrimaryRuntimeFlavor>
     <DefaultPrimaryRuntimeFlavor Condition="'$(TargetArchitecture)' == 's390x'">Mono</DefaultPrimaryRuntimeFlavor>
     <DefaultPrimaryRuntimeFlavor Condition="'$(TargetsLinuxBionic)' == 'true'">Mono</DefaultPrimaryRuntimeFlavor>
     <DefaultPrimaryRuntimeFlavor Condition="'$(TargetsMobile)' == 'true'">Mono</DefaultPrimaryRuntimeFlavor>
```

### Configuration

Docker image: `mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-22.04-cross-riscv64`
`export ROOTFS_DIR=/crossrootfs/riscv64`

### Other information

_No response_

### PR DESCRIPTION
**Pull request by clamp03**
Fix https://github.com/dotnet/runtime/issues/96978 issue

Part of dotnet/runtime#84834
cc @dotnet/samsung

### CODE REVIEW

**Review by jkotas — APPROVED** [review]
LGTM

**Review by tomeksowi — APPROVED** [review]

**Review by am11 — APPROVED** [review]
Looks great!
(one less thing to worry about when adding a new platform support 😅)

File: src/tools/illink/src/linker/Linker.Steps/OutputStep.cs [hunk 1]
```diff
@@ -125,7 +95,7 @@ protected virtual void WriteAssembly (AssemblyDefinition assembly, string direct
 				if (module.IsCrossgened ()) {
 					module.Attributes |= ModuleAttributes.ILOnly;
 					module.Attributes ^= ModuleAttributes.ILLibrary;
-					module.Architecture = CalculateArchitecture (module.Architecture);
+					module.Architecture = 0; // Unknown architecture which ultimately translates to AnyCPU
```
Comment (am11) [inline src/tools/illink/src/linker/Linker.Steps/OutputStep.cs:98]:
```suggestion
					module.Architecture = 1; // MSIL architecture which ultimately translates to AnyCPU
```

**Review by gbalykov — APPROVED** [review]

File: src/tools/illink/src/linker/Linker.Steps/OutputStep.cs (hunk 1 above)
Comment (clamp03) [reply src/tools/illink/src/linker/Linker.Steps/OutputStep.cs]: Thank you so much!

File: src/tools/illink/src/linker/Linker.Steps/OutputStep.cs [hunk 2]
```diff
@@ -125,7 +95,7 @@ protected virtual void WriteAssembly (AssemblyDefinition assembly, string direct
 				if (module.IsCrossgened ()) {
 					module.Attributes |= ModuleAttributes.ILOnly;
 					module.Attributes ^= ModuleAttributes.ILLibrary;
-					module.Architecture = CalculateArchitecture (module.Architecture);
+					module.Architecture = 1; // MSIL architecture which ultimately translates to AnyCPU
```
Comment (am11) [inline src/tools/illink/src/linker/Linker.Steps/OutputStep.cs:98]: To fix the build error:
```suggestion
					module.Architecture = (TargetArchitecture)0x0001; // MSIL architecture which ultimately translates to AnyCPU
```

Comment (clamp03) [reply src/tools/illink/src/linker/Linker.Steps/OutputStep.cs]: Thank you. Now I am out of office so I cannot test. Sorry and thank you for your review.

**Comment by clamp03** [comment]
I will handle failures tomorrow. Thank you.

**Comment by am11** [comment]
I was able to locally repro the illink error we are seeing in the CI; it repros with crossbuild+release config. This patch seem to fix it:

```diff
--- a/src/coreclr/binder/assemblybindercommon.cpp
+++ b/src/coreclr/binder/assemblybindercommon.cpp
@@ -143,8 +143,7 @@ namespace BINDER_SPACE
         }
         else
         {
-            if ((CLRPeKind & peILonly) && !(CLRPeKind & pe32Plus) &&
-                !(CLRPeKind & pe32BitRequired) && dwImageType == IMAGE_FILE_MACHINE_I386)
+            if ((CLRPeKind & peILonly) && !(CLRPeKind & pe32Plus) && !(CLRPeKind & pe32BitRequired))
             {
                 // Processor-agnostic (MSIL)
                 *PeKind = peMSIL;
diff --git a/src/coreclr/vm/peassembly.cpp b/src/coreclr/vm/peassembly.cpp
index c7c618b4852..99ccb8638f9 100644
--- a/src/coreclr/vm/peassembly.cpp
+++ b/src/coreclr/vm/peassembly.cpp
@@ -39,7 +39,7 @@ static void ValidatePEFileMachineType(PEAssembly *pPEAssembly)
     DWORD actualMachineType;
     pPEAssembly->GetPEKindAndMachine(&peKind, &actualMachineType);

-    if (actualMachineType == IMAGE_FILE_MACHINE_I386 && ((peKind & (peILonly | pe32BitRequired)) == peILonly))
+    if ((peKind & (peILonly | pe32BitRequired)) == peILonly)
         return;    // Image is marked CPU-agnostic.

     if (actualMachineType != IMAGE_FILE_MACHINE_NATIVE && actualMachineType != IMAGE_FILE_MACHINE_NATIVE_NI)
```

**Comment by am11** [comment]
Mono has a different behavior than existing coreclr and proposed patch: https://github.com/dotnet/runtime/blob/b7c34460b56bbde2b5bc1e15b877ce475c7f22a6/src/mono/mono/metadata/assembly.c#L499 it ignores the "unknown"  and "msil" cases for non-i386 archs and simply does not set the machine. Unless there is a downside, we can align that to the proposed behavior as well? cc @jkotas, @lambdageek

**Comment by jkotas** [comment]
> I was able to locally repro the illink error we are seeing in the CI; it repros with crossbuild+release config. 

These failures suggests that the fix is not producing proper AnyCPU files. Have you checked that the fix is producing AnyCPU files?

AnyCPU are IMAGE_FILE_MACHINE_I386, COMIMAGE_FLAGS_ILONLY bit set, both COMIMAGE_FLAGS_32BITREQUIRED and COMIMAGE_FLAGS_32BITPREFERRED bits cleared.

I do not think we want to be changing the assembly loader as part of this fix.

**Comment by jkotas** [comment]
> Shouldn't corelib be machine specific?

CoreLib is machine, OS and runtime specific, but this fact is not encoded in the PE header.

The runtime should not require CoreLib to be machine specific.

File: src/tools/illink/src/linker/Linker.Steps/OutputStep.cs [hunk 3]
```diff
@@ -125,7 +95,7 @@ protected virtual void WriteAssembly (AssemblyDefinition assembly, string direct
 				if (module.IsCrossgened ()) {
 					module.Attributes |= ModuleAttributes.ILOnly;
 					module.Attributes ^= ModuleAttributes.ILLibrary;
-					module.Architecture = CalculateArchitecture (module.Architecture);
+					module.Architecture = (TargetArchitecture)0x0001; // MSIL architecture which ultimately translates to AnyCPU
```
Comment (am11) [inline src/tools/illink/src/linker/Linker.Steps/OutputStep.cs:98]:
```suggestion
					module.Architecture = TargetArchitecture.I386; // I386+ILOnly which ultimately translates to AnyCPU
```

Comment (am11) [reply src/tools/illink/src/linker/Linker.Steps/OutputStep.cs:98]: @jkotas this fixes the linker tests.

**Review by sbomer — APPROVED** [review]

**Comment by clamp03** [comment]
@am11 @jkotas Thank you so much!

**Pull request merged**